static/assets/tutorials/ink-workshop/2.4-finished-code.rs

```diff
@@ -4,203 +4,241 @@ use ink_lang as ink;
-			let to_balance = self.balance_of_or_zero(&to);
-			self.balances.insert(to, to_balance + value);
-
-			self.env().emit_event(Transfer {
-				from: Some(from),
-				to: Some(to),
-				value,
-			});
-
-			true
-		}
-
-		fn balance_of_or_zero(&self, owner: &AccountId) -> Balance {
-			*self.balances.get(owner).unwrap_or(&0)
-		}
-
-		fn allowance_of_or_zero(&self, owner: &AccountId, spender: &AccountId) -> Balance {
-			// If you are new to Rust, you may wonder what's the deal with all the asterisks and
-			// ampersends.
-			//
-			// In brief, using `&` if we want to get the address of a value (aka reference of the
-			// value), and using `*` if we have the reference of a value and want to get the value
-			// back (aka dereferencing).
-			// To read more: https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html
-			*self.allowances.get(&(*owner, *spender)).unwrap_or(&0)
-		}
-	}
-
-	#[cfg(test)]
-	mod tests {
-		use super::*;
-
-		use ink_lang as ink;
-
-		#[ink::test]
-		fn new_works() {
-			let contract = Erc20::new(777);
-			assert_eq!(contract.total_supply(), 777);
-		}
-
-		#[ink::test]
-		fn balance_works() {
-			let contract = Erc20::new(100);
-			assert_eq!(contract.total_supply(), 100);
-			assert_eq!(contract.balance_of(AccountId::from([0x1; 32])), 100);
-			assert_eq!(contract.balance_of(AccountId::from([0x0; 32])), 0);
-		}
-
-		#[ink::test]
-		fn transfer_works() {
-			let mut contract = Erc20::new(100);
-			assert_eq!(contract.balance_of(AccountId::from([0x1; 32])), 100);
-			assert!(contract.transfer(AccountId::from([0x0; 32]), 10));
-			assert_eq!(contract.balance_of(AccountId::from([0x0; 32])), 10);
-			assert!(!contract.transfer(AccountId::from([0x0; 32]), 100));
-		}
-
-		#[ink::test]
-		fn transfer_from_works() {
-			let mut contract = Erc20::new(100);
-			assert_eq!(contract.balance_of(AccountId::from([0x1; 32])), 100);
-			contract.approve(AccountId::from([0x1; 32]), 20);
-			contract.transfer_from(AccountId::from([0x1; 32]), AccountId::from([0x0; 32]), 10);
-			assert_eq!(contract.balance_of(AccountId::from([0x0; 32])), 10);
-		}
-
-		#[ink::test]
-		fn allowances_works() {
-			let mut contract = Erc20::new(100);
-			assert_eq!(contract.balance_of(AccountId::from([0x1; 32])), 100);
-			contract.approve(AccountId::from([0x1; 32]), 200);
-			assert_eq!(contract.allowance(AccountId::from([0x1; 32]), AccountId::from([0x1; 32])), 200);
-
-			assert!(contract.transfer_from(AccountId::from([0x1; 32]), AccountId::from([0x0; 32]), 50));
-			assert_eq!(contract.balance_of(AccountId::from([0x0; 32])), 50);
-			assert_eq!(contract.allowance(AccountId::from([0x1; 32]), AccountId::from([0x1; 32])), 150);
-
-			assert!(!contract.transfer_from(AccountId::from([0x1; 32]), AccountId::from([0x0; 32]), 100));
-			assert_eq!(contract.balance_of(AccountId::from([0x0; 32])), 50);
-			assert_eq!(contract.allowance(AccountId::from([0x1; 32]), AccountId::from([0x1; 32])), 150);
-		}
-	}
+    use ink_storage::traits::SpreadAllocate;
```

Comment:
It looks like I changed all this because I switched from tabs to spaces (consistency with the other templates)